Sphinx docs

real-yfprojects opened the pull request:
The docs are almost entirely generated from the `README.md` file.
The documentation uses the `furo` theme with customizations,
adjusting the logo, edit-source button and adding a version selector item to the buttom
of the sidebar reminding of the one by readthedocs.

* .pre-commit-config.yaml (darglint2): Exclude `docs` directory.

* Makefile : Add phonies for tests, lint, docs and clean.

* pyproject.toml : Add `docs` dependency group with sphinx and the used extensions.

* poetry.lock : Autogenerated.

* docs/Makefile : Merge make file autogenerated by `sphinx-quickstart` into this file.

* docs/make.bat : Autogenerated by `sphinx-quickstart`.

* docs/source/.gitignore : Ignore `doctrees` directory generated by sphinx.

* docs/source/conf.py : Configure sphinx.

* docs/source/index.rst : Starting point of doc generation. Generated `index.html`.

* docs/source/changelog.rst : Include `CHANGELOG.md` in docs.

* docs/source/readme.rst : Include `README.md` in docs.

* docs/source/license.rst : Include `LICENSE.txt` in docs.

* docs/source/_static/css/ext-links.css : Add unused css file adding a small marker icon to external links.
	This could be used in the future. I committed it so that this work isn't lost.

* docs/source/_templates/versioning.html : Implement version picker.

* docs/source/_static/css/version-selector.css : Style version picker.

* docs/source/_templates/components/edit-this-page.html : Modify behaviour of `edit-source` button.

* docs/source/_templates/sidebar/brand.html : Style darglint2 logo adding a version number.

real-yfprojects commented:
Next I will implement the github action workflow for generating the docs. I hope to be able to test it on my fork so we can see a live preview. In the meantime you can have a look at them by building them locally. Let me know what you think and what you would like to improve.

Currently the docs content matches the README file of this repo. However I am planning on improving the docs content once this is merged.

akaihola commented:
I see the following Sphinx build warnings:
```
locale_dir docs/source/locales/en/LC_MESSAGES does not exists
README.md:29: (WARNING/2) Document headings start at H2, not H1 [myst.header]
README.md:56: (WARNING/2) Document headings start at H2, not H1 [myst.header]
README.md:70: (WARNING/2) Document headings start at H2, not H1 [myst.header]
README.md:214: (WARNING/2) Document headings start at H2, not H1 [myst.header]
README.md:394: (WARNING/2) Document headings start at H2, not H1 [myst.header]
README.md:404: (WARNING/2) Document headings start at H2, not H1 [myst.header]
README.md:427: (WARNING/2) Document headings start at H2, not H1 [myst.header]
README.md:448: (WARNING/2) Document headings start at H2, not H1 [myst.header]
README.md:498: (WARNING/2) Document headings start at H2, not H1 [myst.header]
README.md:517: (WARNING/2) Document headings start at H2, not H1 [myst.header]
findfont: Font family ['Roboto'] not found. Falling back to DejaVu Sans.
README.md:51: WARNING: 'myst' cross-reference target not found: '/9dogs/darglint/commit/0c8a388715daf4ca8e5a8b76107d7ab8a865ab23#r95373928' [myst.xref_missing]
README.md:51: WARNING: 'myst' cross-reference target not found: '/9dogs/darglint/commit/0c8a388715daf4ca8e5a8b76107d7ab8a865ab23' [myst.xref_missing]
README.md:51: WARNING: 'myst' cross-reference target not found: '/terrencepreilly/darglint' [myst.xref_missing]
```
Docs still bulid just fine, these could be looked at later, too.

akaihola commented:
These warnings:
```bash
README.md:51: WARNING: 'myst' cross-reference target not found: '/9dogs/darglint/commit/0c8a388715daf4ca8e5a8b76107d7ab8a865ab23#r95373928' [myst.xref_missing]
README.md:51: WARNING: 'myst' cross-reference target not found: '/9dogs/darglint/commit/0c8a388715daf4ca8e5a8b76107d7ab8a865ab23' [myst.xref_missing]
README.md:51: WARNING: 'myst' cross-reference target not found: '/terrencepreilly/darglint' [myst.xref_missing]
```
are solved by:
```diff
diff --git a/README.md b/README.md
index 36a73bc..17e89fe 100644
--- a/README.md
+++ b/README.md
@@ -51,9 +51,9 @@ With a few useful bugfixes and pull requests left out of the final 1.8.1 release
 but copied and renamed it to `darglint2` at the request of the original author.
 
 For more background about the fork,
-see also [this discussion](/9dogs/darglint/commit/0c8a388715daf4ca8e5a8b76107d7ab8a865ab23#r95373928)
-attached to commit [0c8a3887](/9dogs/darglint/commit/0c8a388715daf4ca8e5a8b76107d7ab8a865ab23)
-in [@9dogs](https://github.com/9dogs)'s fork of the original [terrencepreilly/darglint](/terrencepreilly/darglint).
+see also [this discussion](https://github.com/9dogs/darglint/commit/0c8a388715daf4ca8e5a8b76107d7ab8a865ab23#r95373928)
+attached to commit [0c8a3887](https://github.com/9dogs/darglint/commit/0c8a388715daf4ca8e5a8b76107d7ab8a865ab23)
+in [@9dogs](https://github.com/9dogs)'s fork of the original [terrencepreilly/darglint](https://github.com/terrencepreilly/darglint).
 
 ## Installation
 
```

akaihola commented:
These warnings:
```py
README.md:29: (WARNING/2) Document headings start at H2, not H1 [myst.header]
README.md:56: (WARNING/2) Document headings start at H2, not H1 [myst.header]
README.md:70: (WARNING/2) Document headings start at H2, not H1 [myst.header]
README.md:214: (WARNING/2) Document headings start at H2, not H1 [myst.header]
README.md:394: (WARNING/2) Document headings start at H2, not H1 [myst.header]
README.md:404: (WARNING/2) Document headings start at H2, not H1 [myst.header]
README.md:427: (WARNING/2) Document headings start at H2, not H1 [myst.header]
README.md:448: (WARNING/2) Document headings start at H2, not H1 [myst.header]
README.md:498: (WARNING/2) Document headings start at H2, not H1 [myst.header]
README.md:517: (WARNING/2) Document headings start at H2, not H1 [myst.header]
```
are solved by:
```diff
diff --git a/README.md b/README.md
index 36a73bc..d62431b 100644
--- a/README.md
+++ b/README.md
@@ -1,7 +1,7 @@
-# Darglint2
-
 <!-- description -->
 
+Darglint2
+
 A functional docstring linter which checks whether a docstring's
 description matches the actual function/method implementation.
 _Darglint2_ expects docstrings to be formatted using the
@@ -14,6 +14,8 @@ would like a feature in _darglint2_.
 
 <!-- end description -->
 
+# Darglint2
+
 **Table of Contents**:
 
 -   [Project Status](#project-status)
```

akaihola commented on README.md:
```diff
@@ -1,5 +1,7 @@
 # Darglint2
 
+<!-- description -->
+
```
```suggestion
<!-- description -->

Darglint2

```

This and the next change below solve these Sphinx warnings:
```py
README.md:29: (WARNING/2) Document headings start at H2, not H1 [myst.header]
README.md:56: (WARNING/2) Document headings start at H2, not H1 [myst.header]
README.md:70: (WARNING/2) Document headings start at H2, not H1 [myst.header]
README.md:214: (WARNING/2) Document headings start at H2, not H1 [myst.header]
README.md:394: (WARNING/2) Document headings start at H2, not H1 [myst.header]
README.md:404: (WARNING/2) Document headings start at H2, not H1 [myst.header]
README.md:427: (WARNING/2) Document headings start at H2, not H1 [myst.header]
README.md:448: (WARNING/2) Document headings start at H2, not H1 [myst.header]
README.md:498: (WARNING/2) Document headings start at H2, not H1 [myst.header]
README.md:517: (WARNING/2) Document headings start at H2, not H1 [myst.header]
```

akaihola commented on README.md:
```diff
@@ -10,6 +12,8 @@ or [Numpy Style Guide](https://numpydoc.readthedocs.io/en/latest/format.html).
 would like a feature in _darglint2_.
 
+<!-- end description -->
+
```
```suggestion

# Darglint2

```

akaihola commented on docs/source/index.rst:
```diff
@@ -0,0 +1,28 @@
+
+.. include:: ../../README.md
+    :parser: myst_parser.sphinx_
+    :start-after: <!-- description -->
```
```suggestion
    :start-after: Darglint2
```
This is one way to solve the H2/H1 issue. The solution depends on how we want the top of `README.md` to look like.

akaihola commented on pyproject.toml:
```diff
@@ -46,6 +46,20 @@ pydocstyle = "^4.0"
 tox    = "^4.4.6"
 
+[tool.poetry.group.docs.dependencies]
+sphinx                           = "^5.3.0"
+sphinx-argparse                  = "^0.3.2"
+sphinx-rtd-theme-github-versions = "^1.1"
+sphinx-rtd-theme                 = "^1.2.0"
+myst-parser                      = "^1.0.0"
+sphinx-notfound-page             = "^0.8.3"
+sphinx-copybutton                = "^0.5.2"
+furo = "^2023.3.27"
+sphinxext-opengraph = "^0.8.2"
+sphinx-autobuild = "^2021.3.14"
+sphinx-design = "^0.4.1"
+sphinx-multiversion = "^0.2.4"
```
`poetry install` now seems to install these dependencies – is that intentional? What if a user just wants to use Darglint2 and doesn't need to build the documentation?

real-yfprojects replied:
It shouldn't. It should only install the deps in the docs group when specifying `--with docs` which can only be used when installing from source afaik.

real-yfprojects replied:
Ah, I had another look into the docs. As I said the dep groups can only be installed from source using poetry. By default all dep groups are installed when installing from source. However you can declare groups to be optional development dependencies. Which groups should be declared as such in your opinion? 

You can always use the `--with` or `--without` flags to select the groups to install.

real-yfprojects replied:
> However you can declare groups to be optional development dependencies. Which groups should be declared as such in your opinion?

@akaihola

akaihola replied:
>[...] optional development dependencies. Which groups should be declared as such in your opinion?

My gut answer is `dev`, `lint`, `test` and `docs`. This way a `git clone ... ; poetry install` creates a minimal footprint and uses minimal bandwidth and time, ending up with a runnable `darglint2`.

But I virtually never use `poetry` myself, so I don't actually know what users of that tool expect.

real-yfprojects replied:
I decided to make lint, test and docs optional since they install additional deps. 

> But I virtually never use `poetry` myself, so I don't actually know what users of that tool expect.

I am also new to poetry but I think installing with poetry is aimed at developers while others could use `pip install git@url`.

real-yfprojects commented:
I am seeing the same warnings as you do. However I ignored them since they are only warnings.
The `myst.header` warnings can be safely ignored since they report exactly what I want. Moreover including the readme in the docs is only a temporary solution imo. 
The warning about the locales can be ignored as well since we don't use localization. Though I fixed it any ways. 
The font warning results from the opengraph extension that renders opengraph preview images. It tries to use the Roboto font which isn't available. However this does not present a problem since it automatically falls back to a font that is installed. There is nothing we can do about this problem since the extension currently doesn't support specifying the font used.

real-yfprojects commented:
I fixed some stuff (but not the README warning) and implemented a script for building the docs for each version. The script is inspired by [sphinx-multiversion](https://github.com/Holzhaus/sphinx-multiversion) which is no longer maintained and doesn't entirely fulfil our needs. I think it was a good decision to implement an own software and I am planning to move the script into its own repo at some point.

akaihola commented on docs/sphinx_polyversion.py:
Was this vendored from another project or created by you?

real-yfprojects replied:
I wrote the entire script.

real-yfprojects replied:
Not yet. I was planning that but I want to setup one project at a time. When the darglint2 repo is all set up and we have time for implementing new features I will come back to the script and make it a reusable python package.

akaihola commented on pyproject.toml:
```diff
@@ -54,11 +54,11 @@ sphinx-rtd-theme                 = "^1.2.0"
+sphinxext-opengraph              = "^0.8.2"
+sphinx-autobuild                 = "^2021.3.14"
+sphinx-design                    = "^0.4.1"
+jinja2                           = "^3.1.2"
```
So you tried with `sphinx-multiversion` first, but ended up using `sphinx_polyversion.py` instead? What was the problem?

real-yfprojects replied:
I don't remember exactly. But the new script has the following advantages:

- Concurrent builds
- More configuration (cmd args)
- Builds in isolated poetry venvs
- Generating global files like the toplevel `index.html` dynamically
- Maintained
- Cleaner and in some aspects also simpler code in my Opinion

real-yfprojects replied:
And I will be able to better integrate the script into github actions - at least in my mind.

real-yfprojects commented:
A preview can be found here: https://real-yfprojects.github.io/darglint2/

real-yfprojects commented:
Doc builds are now even faster because also the isolated poetry environment for each version are cached. I also added docstrings and code comments to the files added in this PR. You shouldn't need to update any settings regarding github pages that should work out of the box once this is merged into master.

akaihola commented on README.md:
Eventually I'd prefer `monospace` to be used in most places where _italics_ is now used – at least for directory paths and file names, command lines and format string attributes.

Current documentation is a mixture of monospace and italics.

The above applies to the change log.

But it's probably best to do this improvement in a separate issue and PR.

real-yfprojects commented:
You have to enable Github Pages in the settings of this repo before the pages can be deployed.

akaihola commented:
> have to enable Github Pages

I did that and triggered a [retry](/akaihola/darglint2/actions/runs/5045340173/jobs/9085281464) of the failed docs build job. Let's see if it succeeds.

real-yfprojects commented:
> I did that and triggered a [retry](/akaihola/darglint2/actions/runs/5045340173/jobs/9085281464) of the failed docs build job. Let's see if it succeeds.

It did. Can you link the docs in the about section like so:
![](https://docs.github.com/assets/cb-137715/mw-1440/images/help/repository/os-repo-with-topics.webp)

akaihola commented:
Linked!
